### DISCUSSION TITLE
How to get Realm from HCL files

### DISCUSSION THREAD

**Discussion by stephenafamo** [ariga/atlas#1431] [Q&A] [answered]
I am trying to directly generate [Bob models](https://github.com/stephenafamo/bob) from the schema defined in the HCL files.

To do that, I am trying to marshal a `schema.Realm` and I can't seem to figure it out.

Here is where I'm at:

```hcl
table "users" {
  schema = schema.example
  column "id" {
    null = false
    type = int
  }
  column "name" {
    null = true
    type = varchar(100)
  }
  primary_key {
    columns = [column.id]
  }
}
```

```go
	parser, err := parseHCLPaths("path/to/directory") // copied this function from atlas
	if err != nil {
		return nil, err
	}

	client := schemahcl.New()
	realm := &schema.Realm{}
	if err := client.Eval(parser, realm, nil); err != nil {
		return nil, err
	}
```

However, this fails with various errors such as:

```
Unknown variable; There is no variable named "schema".
```

And if I comment out the schema line

```
Call to unknown function; There is no function named "varchar".
```

**Comment by rotemtam** [comment, marked as answer]
schemahcl is a fairly low level library. each driver has a method for reading HCL, for example:
https://github.com/ariga/atlas/blob/25fc911e61239957d79cb07f49cf44d0ff420a25/sql/mysql/sqlspec.go#L93

**Reply by stephenafamo** [reply]
Thank you. This is perfect.
